### ISSUE TITLE
Перезапись справочников при обновлении конфигурации клиента

### ISSUE THREAD

**Issue by stasgm** [GoldenSoftwareLtd/gedemin-apps#5]
В бухгалтерии План счетов, в зарплате Види начислений, в фронте розницы Настройка клавиатуры. Все эти справочники у клиентов затираются при обновлении конфигурации через синхронизацию ПИ. Но справочники это не скрипт-функции, которые мы хотим сохранять в едином виде у всех. В настоящем механизме синхронизации это проблема никак не решена. 
Неделей ранее я обновил зп и ок заказчику и потом неделю они звонили по поводу того, что счёт не тот, то сумма не та.
Ваши решения?

**Comment by YRozmysl** [comment]
ты обновляешь с галкой перезаписывать? По-идее, не должно там затирать счета

**Comment by NikolayUkleyko** [comment]
Добрый день !

  Если, конкретные ПИ было установлены на базу данных *впервые *- данные в
справочнике будут затерты независимо от того, была ли проставлена или не
проставлена галка у реквизита *Всегда перезаписывать*.
  Если, конкретные ПИ уже было установлено на базу данных:

   - данные в справочнике *не будут затерты*, если *не проставлена галка* у
   реквизита *Всегда перезаписывать*;
   - данные в справочнике *будут затерты*, если *проставлена галка* у
   реквизита *Всегда перезаписывать*.

  Всего хорошего.




20 марта 2015 г., 9:42 пользователь Yury Rozmysl <notifications@github.com>
написал:

> ты обновляешь с галкой перезаписывать? По-идее, не должно там затирать
> счета
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/GoldenSoftwareLtd/gedemin-apps/issues/5#issuecomment-83939520>
> .
>



-- 
С уважением, Николай.

**Comment by YRozmysl** [comment]
Как я прнял при обновлении конкретно ЗП речь идет именно при обновлении через ПИ, когда уже был ранее установлен пакет ПИ "ЗПиОк"

**Comment by MIchaelShoihet** [comment]
С планом счетов по другому нельзя, т.к. счета есть в эталонной базе, ПИ (или настройка неважно) должны их переписать, что бы установить значения аналитик.  С другой стороны, последующие загрузки по дате должны проверять и не переписывать, если были какие-то изменения.

**Comment by gsbelarus** [comment]
Как задумывалось:

1) Если при загрузке ПИ установлена галка Перезаписывать, то перезапишет всегда
2) Если нет поля EDITIONDATE -- надо эксперементировать. Задумывалось, что в этом случае возьмется дата самого ПИ. Но как сейчас работает код, это надо проверять отдельно.
3) Если есть поле EDITIONDATE -- перезапишет, если в файле новее, чем в БД.

Очевидно, что механизм не идеален. 

Для упомянутых стасом случаев должен действовать такой же механизм, как и для рабочей компании, которая создается при первом подключении к чистой БД.

Т.е. некие параметры, привязанные к текущей БД, должны создаваться, если их нет, и инициализироваться некоторыми стандартными значениями, взятыми из таблиц, из ПИ. В последствии, так как это локально созданные записи, они затираться не будут.

Отрицательный эффект такого нововведения, что если ДЕЙСТВИТЕЛЬНО надо обновить/переписать некоторый параметр из ИП, то сделать это уже будет невозможно без вмешательства оператора.

**Comment by alexandra-gsoftware** [comment]
Может, есть смысл добавить в at_object поле "не перезаписывать"? Тогда, например, возьмем ac_account. В него добавилась новая аналитика, она проставлена в нескольких счетах, счета пересохранены в ПИ. У них стоит галочка не перезаписывать. Тогда значение нового поля должно проставиться, а старые поля - не перезаписываться. Как-то так.

**Comment by gsbelarus** [comment]
Это уже проходили. Так было в настройках. Правильно -- именно создание копий настроек на рабочей БД на основе неких эталонных. В дальнейшем, ты с этой копией делаешь что хочешь, загрузка ПИ копии не затронет.

**Comment by YRozmysl** [comment]
Проблема затирания имеет место во всех простых таблицах в которых нет поля editiondate. При обновлении через ПИ пользовательские изменения всегда затираются, даже если и не стоит галка перезаписывать. Проверно. Возможно, следует сделать в одной из версий exe модифай базы с автоматическим созданием этого поля и заполнить его значением, например, датой модифая?

**Comment by NikolayUkleyko** [comment]
1. На сколько я понял, речь шла о справочнике Виды начислений. В данном справочнике есть поле EDITIONDATE.
  2. Какую таблицу принято считать простой ? Если, при создании таблицы был выбран тип Простая таблица с идентификатором, тогда да. В данной таблице не будет поля EDITIONDATE.
 
![image](https://cloud.githubusercontent.com/assets/11497453/6751825/c5db6fb6-cf19-11e4-99db-17061dae6f3e.png)

  Вообще, не будет ли противоречить идеологии платформы, если обновить структуру базы данных, добавив в простые таблицы с идентификатором поле EDITIONDATE ?

**Comment by YRozmysl** [comment]
Нет, не этот справояник. речь идет об одном из, но других. например "Типы отработанных часов"

**Comment by NikolayUkleyko** [comment]
Тогда, пожалуйста, обратитесь к Администратору с данной просьбой.
  Вообще, не будет ли противоречить идеологии платформы, если обновить структуру базы данных, добавив в простые таблицы с идентификатором поле EDITIONDATE ?

**Comment by YRozmysl** [comment]
Ранее поле EDITIONDATE  чаще всего использовалось для поиска "вредителя", теперь же при работе с ПИ его назначение становится уже более существенным. Не знаю, что и как решать проблему, но решать ее надо обязательно.

**Comment by NikolayUkleyko** [comment]
Пробовал экспериментировать с таблицей Конфигурации книги покупок (USR$BUYINGBOOKCONF). Установка ПИ без установки галки по реквизиту "Всегда перезаписывать" не затерла данные - сработало верно. Хотя, в данной таблице поля EDITIONDATE нет.

**Comment by stasgm** [comment]
А если заглянуть в исходники платформы, то даже можно не экспериментировать)

**Comment by alexandra-gsoftware** [comment]
2  NikolayUkleyko 
мне кажется, там данные просто не входят в ПИ...

**Comment by YRozmysl** [comment]
Вот еще, что. Если в базе загружено, например, ОК - данные  24.10.2014 10:06:25   1.0.0.51 и 
там пользователь делал какие-то изменения. После чего мы обновляем ПИ (без галки перезаписывать), например ОК - данные  19.03.2015 15:14:18   1.0.0.52, то данные во всех простых таблицах затираются данными из 52 версии.
Далее, если мы уже в 52 версии, непосредственно на рабочей базе, делаем изменения какие-то и снова загружаем поверх ОК - данные  19.03.2015 15:14:18   1.0.0.52, то ничего не затирвется. Получается, что в случае, если  нет поля EDITIONDATE проверяется по версии ПИ и FILETIMESTAMP?

**Comment by gsbelarus** [comment]
Да. Если дат нет, то смотрим по номеру версии.

**Comment by NikolayUkleyko** [comment]
Ну, что, Стас, сбылось чудо ?

**Comment by YRozmysl** [comment]
Так, а какое чудо? Тут просто получается, что при  обновлении пакетом. в который входит ПИ.Данные, просто не ставить это ПИ и тогда ничего не затрется.  Либо, если, как я понимаю по-хорошему, надо добавлять поле EDITIONDATE в справочники, где его нету, чтобы при обновлении пакетом проверялось это поле.

**Comment by NikolayUkleyko** [comment]
Раз подняли тему, то надо и реализовать решение по этой теме.
  Проще всего сделать так, чтобы НИКОГДА не перезаписывались данные из ПИ. Необходимо проверять, существуют ли данные в таблицах справочников. Если, да - ничего не перезаписывать. В противном случае эталонные настройки установятся без данных. Справочники будут пустые .
  В результате реквизит "Всегда перезаписывать" при установке данных не будет влиять.
  Жаль, что такой необходимый функционал у нас не реализуется никак. Сколько таких примеров. Переписка есть, а дела - нет.
  :-(

**Comment by MIchaelShoihet** [comment]
К сожалению такое решение аналогично решению, которое есть сейчас, только с другим знаком. Часть данных которые у нас есть в ПИ это какие-то настройки, которые должны переписываться, в случае скажем изменения законодательства, + настройки по счетам на пустую базу то же надо переписывать... Поэтому сам механизм задуман правильно и необходимо просто для начала проверить все ПИ что бы у них в данных правильный флаг стоял и если проблема остается, то решать ее на уровне платформы. Единственный момент - это с планом счетов (он у нас идет в эталоне) и из-за этого возникаю проблемы - если не переписывать данные то загрузив настройки на чистую базу у нас не будут настроены аналитики по счетам, если же загружать, но при загрузки на клиентскую базу могут затереться клиентские настройки....

**Comment by NikolayUkleyko** [comment]
Полностью согласен.
  Но, как ни крути, решить данную проблему можно только на уровне платформы.

**Comment by YRozmysl** [comment]
Наверное, лучше всего  - это в какой-то из версий exe сделать добавление поля EDITIONDATE в справочники, где его нет. И заполнить например, датой модифая или какой-то другой. Чем не решение проблемы?

**Comment by NikolayUkleyko** [comment]
Я больше за проверку значения в только реквизите FILETIMESTAMP, чем по EDITIONDATE. Вариант, если нет поля EDITIONDATE проверяется по версии ПИ и FILETIMESTAMP не годится. Здесь явный приоритет на дату редактирования. Реквизит FILETIMESTAMP носит вторичный характер.
  Пользователь может вызвать на редактирование запись из справочника. Далее ничего не изменяя нажимает кнопку Ok. Вот, пожалуйста, новая дата редактирования. При установке ПИ программа не обновит такие данные, так посчитает, что в БД данные более новые, чем на диске.
  При таких действиях пользователь не сможет поменять значение версии файла ПИ. Считаю завязку на реквизите FILETIMESTAMP более надежной.
  Я про Modify забыл с марта 2015 г. :-)

**Comment by gsbelarus** [comment]
ни EDITIONDATE, ни FILETIMESTAMP проблему не решат. Кто-нибудь на эталоне по щелкает по записям. EDITIONDATE обновится и у всех клиентов перезапишется все.

Такого рода данные должны именно клонироваться в момент установки программы или в момент первого обращения. И дальше жить своей жизнью.

**Comment by alexandra-gsoftware** [comment]
ага, только в коде используются их руиды

**Comment by MIchaelShoihet** [comment]
Может все-таки ориентироваться только на флаг, но сделать его из 3-х значений = Переписывать, Не переписывать, Переписывать только для новой. 3-й флаг будет означать что данные будут переписаны, только если настройка устанавливается первый раз, остальные будут работать как и раньше, для определение 1-го раза добавляем поле которое не переносится с ПИ а заполняется только при загрузке на базу. Тогда если стоит флаг переписывать, то всегда переписываем, если нет то только добавляем если элемента нет, в третьм случае переписываем, только если данное ПИ не было ранее установлено на базу. Этим третьим вариантом решается проблема с бухгалтерскими счетами, на новую базу они установятся как надо, а при обновленнии уже не будут перезаписываться... Это как мысль к размышлению...

**Comment by MIchaelShoihet** [comment]
На дату нельзя ориентироваться, потому что она может измениться не предсказуемо....

**Comment by alexandra-gsoftware** [comment]
>>>Переписывать только для новой. 
А если мы ставим на базу, где нет ПИ, но настроен учет (ставим ПИ первый раз?)

Проблема со счетами имхо решится, если мы их грамотно разнесем по ПИ. Зачем ставить счет? Чтобы использовать его в проводке. Если в ПИ будут только этот счет и эта проводка, то его (ПИ)можно ставить или не ставить (а зачем его ставить, если у клиента уже свои проводки есть?)

**Comment by alexandra-gsoftware** [comment]
Еще можно было бы дать возможность выбирать поля, которые всегда должны обновляться (например, для того, чтоб работали мои электронные док-ты, мне важен код страны, а как она называется, где в дереве стран находится - не важно, и я бы с радостью оставила это так, как настроено у юзера)

**Comment by MIchaelShoihet** [comment]
Грамотная разноска ПИ - это само собой. И в этом случае как-раз и сработает правило, мы ставим первый раз настройку с новым функционалом, если в нее входят счета, значит они входят с проводками и значит с настроенной аналитикой, поэтому будет нормально если мы их заменим....

**Comment by alexandra-gsoftware** [comment]
>>>поэтому будет нормально если мы их заменим....

вот-вот. Наша основная проблема со счетами состоит в том, что они лежат вместе с какими-нить данными, а проводки лежат отдельно. И проводки можно не ставить, но если обновить Торговля - данные, то счета поломаются. С этими счетами темный лес, и я даже не знаю, с какой стороны к ним подойти... https://code.google.com/p/gedemin/issues/detail?id=3558

**Comment by NikolayUkleyko** [comment]
Замечательный вариант с выбором обновляемых полей. Как вариант, по умолчанию все поля обновляются и выбрать кроме каких.

**Comment by alexandra-gsoftware** [comment]
>>>> Как вариант, по умолчанию все поля обновляются

да. В 99,99% случаев никто с полями заморачиваться не будет, зато в 0,01% пригодится.

Еще, если у нас есть галочка перезаписывать\не перезаписывать. Я бы предложила при добавлении объекта в ПИ ставить ее по умолчанию в зависимости от класса (т.е. хранимую процедуру мы перезаписываем, а attr user defined - нет....), а дальше пусть разработчик решает.

**Comment by NikolayUkleyko** [comment]
Много-много сообщений: обновить или нет.

**Comment by YRozmysl** [comment]
Может, все же  как-то вернемся к проблеме? Обновили на очередной базе базе ПИ, и пере затерлись справочники "Должностей", "Видов отпуска", "Типов больничных" и т.д. А просто так, тупо, ставить весь пакет за исключением этих ПИ я не не могу, так как там могут появится новые данные, руиды которых используются в расчетах.

**Comment by NikolayUkleyko** [comment]
Обсудим тему.
1. ПИ устанавливается на базу данных, на которой уже было установлено это ПИ ?
2. При установке ПИ в параметрах не проставлена ли галка по реквизиту **Всегда перезаписывать** ? Для избежания затирания данных необходимо снять галку с такого реквизита.
3. Здесь уже многое было сказано, как решить данную проблему: проверка по RUID, по дате изменения файла ПИ и т. д. В итоге проблема осталась. Нельзя ли предусмотреть **точную** проверку по содержимому в справочниках (по символам, по тексту, по цифрам) ?

**Comment by YRozmysl** [comment]
Вопрос решается. https://github.com/GoldenSoftwareLtd/gedemin/issues/3656  Просто надо добавить системные поля в нужные таблицы. По данной ссылке, справочники, которые касаются ЗПиОК только.

**Comment by NikolayUkleyko** [comment]
Хорошо.
  Тогда, эту тему может закроем ?
